app/src/main/native/cpp/looper_thread.cpp

```diff
@@ -1,81 +1,28 @@
 #include "looper_thread.hpp"
 
-namespace engine {
-namespace android {
+#include <android/looper.h>
 
-LooperThread::LooperThread(): thread(std::thread(&LooperThread::eventLoop, this)) {
-  LOGI("LooperThread constructor");
-}
+#include <memory>
 
-LooperThread::~LooperThread() {
-  LOGI("LooperThread destructor, sent destroy task, waiting...");
-  scheduleTask([this] {
-    ALooper_removeFd(aLooper, fds[PIPE_OUT]);
-    if (close(fds[PIPE_IN]) || close(fds[PIPE_OUT])) {
-      throw std::runtime_error("Failed to close file descriptor!");
-    }
-    // explicit wake here in order to unblock ALooper_pollAll
-    ALooper_wake(aLooper);
-    ALooper_release(aLooper);
-    LOGI("Looper is released!");
-  });
-  thread.join();
-  LOGI("LooperThread destructor, thread killed!");
-  taskMap.clear();
-}
+#include "run_loop.hpp"
+#include "util.hpp"
 
-void LooperThread::eventLoop() {
-  // in order to use AChoreographer for effective rendering
-  // and allow scheduling events in general we prepare and acquire the ALooper
-  aLooper = ALooper_prepare(0);
-  assert(aLooper);
-  ALooper_acquire(aLooper);
-  if (pipe2(fds, O_CLOEXEC)) {
-    throw std::runtime_error("Failed to create pipe needed for the ALooper");
-  }
-  // TODO understand if this is really required
-  if (fcntl(fds[PIPE_OUT], F_SETFL, O_NONBLOCK)) {
-    throw std::runtime_error("Failed to set pipe read end non-blocking.");
-  }
-  auto ret = ALooper_addFd(aLooper, fds[PIPE_OUT], ALOOPER_POLL_CALLBACK,
-                           ALOOPER_EVENT_INPUT, looperCallback, this);
-  if (ret != 1) {
-    throw std::runtime_error("Failed to add file descriptor to Looper.");
-  }
-
-  int outFd, outEvents;
-  char *outData = nullptr;
+namespace engine {
+namespace android {
 
-  // using negative timeout and block poll forever
-  // we will be using poll callbacks to schedule events and not make use of ALooper_wake() at all unless we want to exit the thread
-  ALooper_pollAll(-1, &outFd, &outEvents, reinterpret_cast<void**>(&outData));
+LooperThread::LooperThread() : runloop_(ALooper_prepare(0)) {
+    LOGI("LooperThread constructor");
+    thread_ = std::thread([&]() { runloop_.run(); });
```

Comment:
as a side note: perhaps we may want allow setting priority as well here? also I wonder if there's difference between `std::thread` and `pthread`

Reply:
> as a side note: perhaps we may want allow setting priority as well here? also I wonder if there's difference between `std::thread` and `pthread`

Yeah. We definitely can add priority support.

> `std::thread` and `pthread`

The first one is coming from the C++ standard library (available since C++11) and the second one is `POSIX Threads` library that was used mainly on Linux platforms before C++11.